rmc/src/handler.rs

```diff
@@ -0,0 +1,122 @@
+            maybe_multisigned = self.handle_signed_hash(signed_hash.clone());
+        }
+        (signed_hash, maybe_multisigned)
+    }
+    /// Update the internal state with the signed hash. If the hash is incorrectly signed then
+    /// [`Error::BadSignature`] is returned. If Adding this signature completes a multisignature
+    /// then `Ok(multisigned)` is returned. Otherwise `Ok(None)` is returned.
+    pub fn on_signed_hash(
+        &mut self,
+        unchecked: UncheckedSigned<Indexed<H>, MK::Signature>,
+    ) -> Result<Option<Multisigned<H, MK>>, Error> {
+        let signed_hash = unchecked
+            .check(&self.keychain)
+            .map_err(|_| Error::BadSignature)?;
+        Ok(
+            match self.already_completed(signed_hash.as_signable().as_signable()) {
+                true => None,
+                false => self.handle_signed_hash(signed_hash),
+            },
+        )
+    }
+
+    fn handle_signed_hash(&mut self, signed: Signed<Indexed<H>, MK>) -> Option<Multisigned<H, MK>> {
+        let hash = signed.as_signable().as_signable().clone();
+        let new_state = match self.hash_states.remove(&hash) {
+            None => signed.into_partially_multisigned(&self.keychain),
+            Some(partial) => partial.add_signature(signed, &self.keychain),
+        };
+        match new_state {
+            PartiallyMultisigned::Complete { multisigned } => {
+                self.hash_states.insert(
+                    hash,
+                    PartiallyMultisigned::Complete {
+                        multisigned: multisigned.clone(),
+                    },
+                );
+                Some(multisigned)
+            }
+            incomplete => {
+                self.hash_states.insert(hash, incomplete);
+                None
+            }
+        }
+    }
+
+    /// Update the internal state with the finished multisigned hash. If the hash is incorrectly
+    /// signed then [`Error::BadMultisignature`] is returned. Otherwise `multisigned` is returned,
+    /// unless the multisignature got completed earlier.
+    pub fn on_multisigned_hash(
+        &mut self,
+        unchecked: UncheckedSigned<H, MK::PartialMultisignature>,
+    ) -> Result<Option<Multisigned<H, MK>>, Error> {
+        if self.already_completed(unchecked.as_signable()) {
+            return Ok(None);
+        }
+
+        let multisigned = unchecked
+            .check_multi(&self.keychain)
+            .map_err(|_| Error::BadMultisignature)?;
+        self.hash_states.insert(
+            multisigned.as_signable().clone(),
+            PartiallyMultisigned::Complete {
+                multisigned: multisigned.clone(),
+            },
+        );
+        Ok(Some(multisigned))
+    }
+
+    fn already_completed(&self, hash: &H) -> bool {
+        matches!(
+            self.hash_states.get(hash),
+            Some(PartiallyMultisigned::Complete { .. })
+        )
+    }
+}
```

Comment:
It would be nice to have tests here – in particular just deasyncing the scenarios from `service.rs` tests should be fine and not too much effort. Although it would also be great if you checked for the specific situation in which the test were failing, so that `start_rmc` returns a multisig.

Reply:
ok, I'll try to come up with some tests